### PR TITLE
[commands] Improve code structure for registeration of commands

### LINKED ISSUE #213: [Commands] Improve code structure

### PR DESCRIPTION
**Pull request by ikfir**
### ❓ Type of change

- [x] 📖 Documentation (updates to the documentation)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (adds functionality)

### 📚 Description

Improve code structure for registering a command.
Added `CollectionHelpers#putAllIfAbsent(Map<K, V>, Map<K, V>)` to add all elements from one map to another if key isn't present.

Closes #213

### 🧪 How Has This Been Tested?
Unit testing

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

## Summary by CodeRabbit

- **New Features**
  - Enhanced command handling with improved serialization and support for named command arguments.
  - Introduced a new method for conditionally adding map entries, `putAllIfAbsent`.
  - Added methods in `ParserFactory` and `RequirementFactory` for improved annotation handling.

- **Bug Fixes**
  - Improved error handling in command tests for more precise reporting.

- **Refactor**
  - Streamlined the `RegisteredCommand` class for better readability and maintainability.

- **Documentation**
  - Updated changelog to reflect recent enhancements and modifications.

### CODE REVIEW

File: commands/src/main/java/net/apartium/cocoabeans/commands/requirements/ArgumentRequirementFactory.java [hunk 1]
```diff
@@ -10,11 +10,44 @@
...
+                return constructor.newInstance();
+
+            if (constructor.getParameters().length == 1 && constructor.getParameterTypes()[0].equals(CommandManager.class))
+                return constructor.newInstance(commandManager);
+
+            return null;
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+            return null;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ikfir) [reply commands/src/main/java/net/apartium/cocoabeans/commands/requirements/ArgumentRequirementFactory.java:43]: there is cache in command manager itself

File: commands/src/main/java/net/apartium/cocoabeans/commands/RegisteredCommand.java [hunk 2]
```diff
@@ -111,57 +92,98 @@ public void addNode(CommandNode node) {
...
+                        ),
+                        HANDLE_EXCEPTION_VARIANT_COMPARATOR
+                );
+            } catch (IllegalAccessException e) {
+                Dispensers.dispense(e);
+                return;
+            }
+        }
+
+    }
```
[the remark this answers is not in the document]
Comment (ikfir) [reply commands/src/main/java/net/apartium/cocoabeans/commands/RegisteredCommand.java:152]: no

File: commands/src/main/java/net/apartium/cocoabeans/commands/parsers/ParserFactory.java [hunk 3]
```diff
@@ -17,6 +18,42 @@
 @ApiStatus.AvailableSince("0.0.30")
 public interface ParserFactory {
 
+    /**
+     * Get the class of the parser factory from an annotation
+     * @param annotation The annotation that has a CommandParserFactory
+     * @return The class of the parser factory
+     */
+    @ApiStatus.AvailableSince("0.0.37")
+    static Class<? extends ParserFactory> getParserFactoryClass(Annotation annotation) {
```
Comment (liorsl) [inline commands/src/main/java/net/apartium/cocoabeans/commands/parsers/ParserFactory.java:27]: why do we need this method, it doesn't look like you use it

File: commands/src/main/java/net/apartium/cocoabeans/commands/parsers/ParserFactory.java [hunk 4]
```diff
@@ -17,6 +18,42 @@
...
+        if (commandParserFactory == null)
+            return null;
+
+        if (!commandParserFactory.scope().isClass() && onlyClassParser)
+            return null;
+
+        try {
+            return commandParserFactory.value().getConstructor().newInstance();
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+            throw new IllegalStateException("Failed to instantiate parser factory: " + commandParserFactory.value(), e);
```
Comment (liorsl) [inline commands/src/main/java/net/apartium/cocoabeans/commands/parsers/ParserFactory.java:53]: IllegalStateException is not really good here, what about RuntimeException?

File: commands/src/main/java/net/apartium/cocoabeans/commands/requirements/ArgumentRequirementFactory.java [hunk 5]
```diff
@@ -10,11 +10,70 @@
...
+
+            if (constructor.getParameters().length == 1 && constructor.getParameterTypes()[0].equals(CommandManager.class)) {
+                if (commandManager == null)
+                    throw new IllegalArgumentException("Command manager cannot be null");
+                return constructor.newInstance(commandManager);
+            }
+
+            return null;
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+            throw new IllegalStateException("Failed to instantiate argument requirement factory: " + clazz, e);
```
Comment (liorsl) [inline commands/src/main/java/net/apartium/cocoabeans/commands/requirements/ArgumentRequirementFactory.java:67]: IllegalStateException is not really good here, what about RuntimeException?

File: commands/src/main/java/net/apartium/cocoabeans/commands/requirements/RequirementFactory.java [hunk 6]
```diff
@@ -11,10 +11,56 @@
...
+
+        Class<? extends RequirementFactory> requirementFactoryClass = getRequirementFactoryClass(annotation);
+
+        if (requirementFactoryClass == null)
+            return null;
+
+        try {
+            return requirementFactoryClass.getConstructor().newInstance();
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+            throw new IllegalStateException("Failed to instantiate requirement factory: " + requirementFactoryClass, e);
```
Comment (liorsl) [inline commands/src/main/java/net/apartium/cocoabeans/commands/requirements/RequirementFactory.java:54]: IllegalStateException is not really good here, what about RuntimeException?

File: common/src/main/java/net/apartium/cocoabeans/CollectionHelpers.java [hunk 7]
```diff
@@ -157,6 +158,20 @@ public static <E> void addElementSorted(List<E> list, E element, Comparator<E> c
...
 
+    /**
+     * Put all entries from other map into map if key is not present
+     * @param map map to put into
+     * @param other map to get entries from
+     * @param <K> key type
+     * @param <V> value type
+     */
+    @ApiStatus.AvailableSince("0.0.37")
+    public static <K, V> void putAllIfAbsent(Map<K, V> map, Map<K, V> other) {
```
Comment (liorsl) [inline common/src/main/java/net/apartium/cocoabeans/CollectionHelpers.java:169]: what about mergeInto?

**Review by liorsl — CHANGES_REQUESTED** [review]

File: commands/src/main/java/net/apartium/cocoabeans/commands/parsers/ParserFactory.java (hunk 3 above)
Comment (ikfir) [reply commands/src/main/java/net/apartium/cocoabeans/commands/parsers/ParserFactory.java:27]: it does

File: common/src/main/java/net/apartium/cocoabeans/CollectionHelpers.java (hunk 7 above)
Comment (ikfir) [reply common/src/main/java/net/apartium/cocoabeans/CollectionHelpers.java:169]: change the name into `mergeInto`

**Review by liorsl — APPROVED** [review]

**Pull request merged**